src/metadata/selectors.js

```diff
@@ -1,225 +1,181 @@
-    // compute the  combination of category-catgoryOption- this basically computes the categoryCombinations
-    // result is a matrix of categoryOptions for each column in the categoryCombo
-    const optionCombinationOrder = cartesian(allOptionsLists)
-    return optionCombinationOrder.map((opts) =>
-        getCoCByCategoryOptions(metadata, catComboId, opts)
-    )
-}
+
+export const selectDataElementsByCatComboInOrder = createSelector(
+    (_, dataElements) => dataElements,
+    selectCategoryCombos,
+    (dataElements, categoryCombos) => {
+        const grouped = []
+
+        dataElements.forEach((de, i) => {
+            const previousDe = dataElements[i - 1]
+            const lastGroup = grouped[grouped.length - 1]
+            const matchingId =
+                previousDe?.categoryCombo.id === de.categoryCombo.id
+
+            if (!lastGroup || !matchingId) {
+                grouped.push({
+                    dataElements: [de],
+                    categoryCombo: categoryCombos[de.categoryCombo.id],
+                })
+            } else {
+                lastGroup.dataElements.push(de)
+            }
+        })
+
+        return grouped
+    }
+)
 
 /**
- * Tries to find the categoryOptionCombo with the given categoryOptions within
- * a category combination.
- * @param {} metadata
- * @param {*} catComboId
- * @param {*} categoryOptionIds
- * @returns
+ * Returns an array of objects with dataElements and their associated categoryCombos. Unlike
+ * selectDataElementsByCatComboInOrder, this selector will group all dataElements with the
+ * same categoryComboId together.
  */
-export const getCoCByCategoryOptions = (
-    metadata,
-    catComboId,
-    categoryOptionIds
-) => {
-    const cocs = getCategoryOptionCombosByCategoryComboId(metadata, catComboId)
-    const sorted = [...categoryOptionIds].sort() //sort to check for equality
-
-    for (const coc of cocs) {
-        const sortedCatOptions = [...coc.categoryOptions].sort()
-        if (
-            sorted.length === sortedCatOptions.length &&
-            sortedCatOptions.every((id, ind) => id === sorted[ind])
-        ) {
-            return coc
+
+export const selectDataElementsByCatCombo = createSelector(
+    (_, dataElements) => dataElements,
+    selectCategoryCombos,
+    (dataElements, categoryCombos) => {
+        // Group dataElements by their categoryCombo id
+        const grouped = groupBy(dataElements, (de) => de.categoryCombo.id)
+
+        // Map to an array and include the associated categoryCombo
+        return Object.entries(grouped).map(
+            ([categoryComboId, dataElements]) => ({
+                dataElements,
+                categoryCombo: categoryCombos[categoryComboId],
+            })
+        )
+    }
+)
+
+/**
+ * Tries to find the categoryOptionCombo with the given categoryOptions within a category
+ * combination.
+ */
+
+export const selectCoCByCategoryOptions = createCachedSelector(
```

Comment:
The signature of the functions are not really obvious anymore. So I think we should make an effort to comment with JSDOC or something.

Reply:
Yeah that sounds good.

Reply:
I'll resolve it, let me know if there's still stuff that needs to be addressed!